### PR TITLE
Add `mindir` format support.

### PR DESCRIPTION
**Pull request by Ethan-Chen-plus**
🤗👋Hi! We have added our `.mindir` format. Please review our commit.

### CODE REVIEW

**Comment by lutzroeder** [comment]
@Ethan-Chen-plus can you provide some context:
* What is mindir, what problem is it solving and for how many users?
* Why is a new format needed given there are many similar formats?
* Create a script in `./tools` following the patterns used for other frameworks to recreate the `-proto` file.
* Add a test to `./test/models.json` and share an index of 50+ existing files in this format.

**Comment by Ethan-Chen-plus** [comment]
```ps
PS I:\netron> npm run validate
eslint
pylint
test
gguf/phi-2.Q2_K.gguf
onnx/candy.onnx
keras/1151.4.keras
coreml/Exermote.mlmodel
pytorch/alexnet.ptl
pytorch/DCGAN2.pt
tf/conv-layers.pb.zip
tflite/mobilenet_v1_0.75_160_quantized.tflite
tflite/squeezenet.tflite
```
I have updated my code, please review.
@lutzroeder

**Comment by lutzroeder** [comment]
@Ethan-Chen-plus can you provide some context:

* What is mindir, what problem is it solving and for how many users?
* Why is a new format needed given there are many similar formats?
* Please rebase to latest `main`.
* The `./tools` scripts should generate regenerate to files from the original `.proto`. When changing the format or generator for `-proto.js` or `-metadata.json` these files should automatically get regenerated from the original source when running `./tools/mindir sync schema metadata`.

**Comment by Ethan-Chen-plus** [comment]
@lutzroeder Thank you again and we have updated our code. And let me introduce some background infomation:

- What is mindir, what problem is it solving and for how many users?
MindIR (MindSpore IR) is a graph-based functional Intermediate Representation (IR) format developed for the MindSpore framework by Huawei. Its primary purpose is to define the logical structure of networks and the attributes of operators, facilitating the decoupling of model files from hardware platforms through a unified IR. This supports a train-once, deploy-multiple-times paradigm. A key advantage of the MindIR format is its ability to store network structures and weight parameter values, enabling deployment for inference tasks on cloud-based Serving and edge-oriented Lite platforms. The same MindIR file can be deployed across various hardware types, including Ascend AI processors, GPUs, and CPUs.

The user base for the MindIR format primarily comprises researchers, developers, and enterprises needing to deploy AI models across multiple hardware platforms. Given its compatibility with diverse hardware including Ascend AI processors, GPUs, and CPUs, MindIR is particularly well-suited for applications that require deployment both in the cloud and on edge devices. For instance, MindIR can be utilized in Huawei's Atlas 200/300/500 inference products, which are specifically designed for edge AI processing scenarios.

-Why is a new format needed given there are many similar formats?

Compared to other similar IR formats such as ONNX and AIR, MindIR presents unique advantages and application scenarios. ONNX (Open Neural Network Exchange) is a universal format designed for the expression of machine learning models, primarily used for model transfer between different frameworks or on inference engines like TensorRT. AIR (Ascend Intermediate Representation), defined by Huawei, is an open file format tailored for machine learning to better accommodate Huawei AI processors, typically used for inference tasks on Ascend 310. MindIR, on the other hand, is a functional IR based on graph representation. It not only defines an extensible graph structure and operator IR representations but also eliminates model discrepancies across different backends. It is generally used for cross-platform inference tasks, such as executing inference on Ascend 310, GPUs, and MindSpore Lite for models trained on Ascend 910.

**Comment by Ethan-Chen-plus** [comment]
@lutzroeder Hello.👋 I noticed that eslint failed. Could you please tell me which config file eslint is using? I'll make the necessary fixes based on that config file. Thank you!

**Comment by Ethan-Chen-plus** [comment]
Requires approval, thank you🤗

**Comment by Ethan-Chen-plus** [comment]
Workflow requires approval, thank you again🤗

File: tools/mindir [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+clean() {
+    echo "mind_ir clean"
+    rm -rf "./third_party/source/mindspore"
+}
+
+sync() {
+    echo "mind_ir sync"
+    mkdir -p "./third_party/source/mindspore/mindspore/mindir/schema/"
+    curl --silent --location --output "./third_party/source/mindspore/mindspore/mindir/schema/mind_ir-metadata.json" "https://raw.githubusercontent.com/Ethan-Chen-plus/mindir/main/mind_ir-metadata.json"
```
Comment (lutzroeder) [inline tools/mindir:14]: The metadata file should not be downloaded from an external location and instead be updated in place as part of `./tools/mindir metadata`.

File: source/view.js [hunk 2]
```diff
@@ -5455,6 +5455,34 @@ view.ModelFactoryService = class {
         this.register('./modular', ['.maxviz']);
         this.register('./cambricon', ['.cambricon']);
         this.register('./weka', ['.model']);
+        this.register('./mslite', ['.ms']);
```
Comment (lutzroeder) [inline source/view.js:5458]: * Resolve incorrect merge

File: source/mind_ir-metadata.json [hunk 3]
```diff
@@ -0,0 +1,406 @@
+[
```
Comment (lutzroeder) [inline source/mind_ir-metadata.json:1]: * Rename to `mindir-metadata.json`

File: source/mind_ir.js [hunk 4]
```diff
@@ -0,0 +1,755 @@
+/**
```
Comment (lutzroeder) [inline source/mind_ir.js:1]: Rename to `mindir.js`.

File: source/mind_ir.js [hunk 5]
```diff
@@ -0,0 +1,755 @@
...
+ * @fileoverview
+ * This file contains the implementation of the mind_ir namespace, including the ModelFactory, Model, and Graph classes.
+ * The ModelFactory class is responsible for matching and opening mind_ir models.
+ * The Model class represents a mind_ir model and contains information about the model's format, name, and graphs.
+ * The Graph class represents a graph within a mind_ir model and contains information about the graph's inputs, outputs, and nodes.
+ */
+export const protobuf = {};
+const mind_ir = {
```
Comment (lutzroeder) [inline source/mind_ir.js:9]: * Rename to `const mindir = {};`.
* Follow pattern in other files that do not predefine empty classes.

File: source/mind_ir.js [hunk 6]
```diff
@@ -0,0 +1,755 @@
+/**
+ * @fileoverview
+ * This file contains the implementation of the mind_ir namespace, including the ModelFactory, Model, and Graph classes.
+ * The ModelFactory class is responsible for matching and opening mind_ir models.
+ * The Model class represents a mind_ir model and contains information about the model's format, name, and graphs.
+ * The Graph class represents a graph within a mind_ir model and contains information about the graph's inputs, outputs, and nodes.
+ */
+export const protobuf = {};
```
Comment (lutzroeder) [inline source/mind_ir.js:8]: * Remove, this should not be needed.

File: source/mind_ir.js [hunk 7]
```diff
@@ -0,0 +1,755 @@
...
+};
+//const mind_ir = mind_ir || {};
+
+mind_ir.ModelFactory = class {
+
+    match(context) {
+        const identifier = context.identifier;
+        const extension = identifier.split('.').pop().toLowerCase();
+        if (extension === "mindir") {
+            return 'mind_ir';
```
Comment (lutzroeder) [inline source/mind_ir.js:37]: * This has changed in more recent versions. Sync to latest.

File: source/mind_ir.js [hunk 8]
```diff
@@ -0,0 +1,755 @@
...
+    get outputs() {
+        return this._outputs;
+    }
+
+    get nodes() {
+        return this._nodes;
+    }
+};
+
+mind_ir.Parameter = class {
```
Comment (lutzroeder) [inline source/mind_ir.js:192]: Renamed in latest. `Parameter` to `Argument` and `Argument` to `Value`.

File: source/mind_ir.js [hunk 9]
```diff
@@ -0,0 +1,755 @@
...
+        }
+        return '';
+    }
+};
+
+mind_ir.Error = class extends Error {
+
+    constructor(message) {
+        super(message);
+        this.name = 'SnapML Model Load Error.';
```
Comment (lutzroeder) [inline source/mind_ir.js:749]: Fix error message.

File: source/mind_ir.js [hunk 10]
```diff
@@ -0,0 +1,755 @@
...
+    SYMBOL: 35,
+    TYPE_NULL: 36,
+    MAP_TENSOR: 37
+};
+
+mind_ir.Attribute = class {
+
+    constructor(context, attribute) {
+
+        this._name = attribute.name;
```
Comment (lutzroeder) [inline source/mind_ir.js:430]: `Attribute` has been removed in latest and replaced with using `Argument` to represent attribtues.

**Review by lutzroeder — CHANGES_REQUESTED** [review]
> Compared to other similar IR formats such as ...

This format looks very similar to ONNX. Instead of defining a new format, can this be accomplished by using `metadata` extensions in ONNX to augment with additional information.

There is also an existing MindSpore Lite format and a dozen other formats that look similar in principle. Consider a path that leverages existing formats or standards instead of creating another duplicate format that needs to be maintained.

**Comment by lutzroeder** [comment]
Closing as stale. Please comment with additional information if the pull request should be reopened.

**Pull request closed without merging**